lib/awful/screen.lua

```diff
@@ -234,6 +235,63 @@ function screen.object.set_padding(self, padding)
     end
 end
 
+--- A list of outputs for this screen with their size in mm.
+--
+-- This information is only available when RANDR is used to manage the screens.
+--
+-- Please note that the table content may vary.
+--
+-- An easy way to check if a screen is the laptop screen is usually:
+--
+--    if s.outputs["LVDS-1"] then
+--        -- do something
+--    end
+--
+-- **Signal:**
+--
+--  * *property::outputs*
+--
+-- **Immutable:** true
+-- @property outputs
+-- @param table
+-- @tfield table table.name A table with the screen name as key (like `eDP1` on a laptop)
+-- @tfield integer table.mm_width The screen physical width
+-- @tfield integer table.mm_height The screen physical height
+-- @tfield integer table.name The screen physical height
+-- @tfield integer table.viewport_id The identifier of the viewport this output
+--  corresponds to.
+
+function screen.object.get_outputs(s)
+    local ret = {}
+
+    local outputs = s._custom_outputs
+        or (s.data.viewport and s.data.viewport.outputs or s._outputs)
```

Comment:
Do I read this correctly: There are three different ways how the outputs of a screen can be saved? That seems a bit excessive to me, but I have no clue what I am looking at currently anyway...

Reply:
### First way

The user or a module sets the outputs manually for unknown reasons:

    my_screen.outputs = {foo = {mm_width =1, mm_height = 1}}

### Second way

Screens managed by CAPI have use the outputs provided by CAPI

### Third way

Screens managed by Lua provide the output stored the semi private `viewport` table. These outputs objects have previously been extended with a `dpi` and diagonal size. In the "new" screen mode, DPIs are per outputs to better handle some clone mode DPI problems previously handled, but unconfigurable.